File: Homework #1/Solutions/python/mygraph.py

""" mygraph.py - Problem #1

    Matthew Romano
    eecs592
"""

import csv


class Node:
    """ Node class """
    def __init__(self, name, key):
        self.index = key
        self.name = name
        self.neighbors = {}

    def add_neighbor(self, neighbor, weight=0):
        """ adds a neighbor and associated edge weight """
        self.neighbors[neighbor] = weight

    def get_neighbors(self):
        """ retrieves list of neighbor nodes """
        return self.neighbors.keys()


class Graph:
    """ Graph class """
    def __init__(self):
        self.nodes = {}
        self.num_nodes = 0
        self.num_edges = 0

    def add_node(self, node_name):
        """ adds a node to the graph """
        if node_name not in self.nodes:
            new_node = Node(node_name, self.num_nodes)
            self.nodes[node_name] = new_node
            self.num_nodes += 1

    def add_edge(self, node1_name, node2_name, edge_weight):
        """ adds an edge to the graph """
        if node1_name not in self.nodes:
            print(node1_name, " is not in the graph")
            return
        if node2_name not in self.nodes:
            print(node2_name, " is not in the graph")
            return
        self.nodes[node1_name].add_neighbor(self.nodes[node2_name], edge_weight)
        self.nodes[node2_name].add_neighbor(self.nodes[node1_name], edge_weight)
        self.num_edges += 1


def main():
    """ main function """
    # 1) Instantiate a Graph
    my_graph = Graph()

    # 2) Read in the data to populate the graph
    with open('graph.txt', 'r') as infile:
        csvreader = csv.reader(infile)
        for line in csvreader:
            node1_name = line[0]
            node2_name = line[1][1:]
            edge_weight = float(line[2][1:])

            my_graph.add_node(node1_name)
            my_graph.add_node(node2_name)
            my_graph.add_edge(node1_name, node2_name, edge_weight)

    # 3) Return the number of Vertices and Edges to the terminal
    print("# of Nodes = ", my_graph.num_nodes)
    print("# of Edges = ", my_graph.num_edges)


if __name__ == "__main__":
    main()

File: Homework #1/Solutions/cpp/mygraph.cpp
/*
 Matthew Romano
 EECS 592
 Problem #3 - mygraph.cpp
*/

#include <iostream>
#include <fstream>        // file input/output
#include <string>         // string
#include <vector>         // vector
#include <unordered_map>  // unordered_map

// Function Declarations
void addVertex(std::string newVertexLabel);
void addEdge(int vertex1, int vertex2, int weight);
int returnNumVertices();
int returnNumEdges();
int returnNumEdgesIter();

// Edge Class
class Edge {
 public:
  Edge() : source(-1), dest(-1), weight(-1) {}
  Edge(int s, int d, int w) : source(s), dest(d), weight(w) {}

  int source;
  int dest;
  int weight;
};


// Graph Variables
typedef std::unordered_map<int, Edge> EdgeMap;  // Need multiple instances
std::unordered_map<int, EdgeMap> vertexToEdgeMap;
std::unordered_map<std::string, int> vertexLabelToVertex;
int vertexCounter = 0;
int edgeCounter = 0;



int main() {
    // File Operation Variables
    std::ifstream inFile("graph.txt");
    std::string line, vertex1, vertex2, weight;
    char delim = 44;  // comma
    size_t p, q;

    // 1) Read in the data
    if (inFile.is_open()) {
        while (getline(inFile, line)) {
            p = line.find_first_of(delim);
            q = line.find_last_of(delim);

            vertex1 = line.substr(0, p);
            vertex2 = line.substr(p+2, q-(p+2));
            weight = line.substr(q+2);

            // Add Vertices and Edges to Graph
            addVertex(vertex1);
            addVertex(vertex2);
            addEdge(vertexLabelToVertex[vertex1], vertexLabelToVertex[vertex2],
                                                                 stoi(weight));
        }
        inFile.close();
    } else {
        std::cout << "Unable to open file";
    }

    // 2) Report # of vertices and edges
    std::cout << "# of Vertices = " << returnNumVertices() << std::endl;
    std::cout << "# of Edges = " << returnNumEdges() << std::endl;


    return 0;
}


/*
*	addVertex (checks whether or not vertex is in the map, adds it if not)
*/
void addVertex(std::string newVertexLabel) {
    std::unordered_map<std::string, int>::iterator it;
    it = vertexLabelToVertex.find(newVertexLabel);
    if (it == vertexLabelToVertex.end()) {
        vertexLabelToVertex[newVertexLabel] = vertexCounter;
        vertexCounter++;
    }
}

/*
*	addEdge
*/
void addEdge(int vertex1, int vertex2, int weight) {
    std::unordered_map<int, EdgeMap>::iterator it;
    it = vertexToEdgeMap.find(vertex1);
    if (it == vertexToEdgeMap.end()) {
        vertexToEdgeMap[vertex1] = EdgeMap();
    }

    vertexToEdgeMap[vertex1][vertex2] = Edge(vertex1, vertex2, weight);
    edgeCounter++;
}

/*
*	returnNumVertices
*/
int returnNumVertices() {
    return vertexCounter;
}

/*
*	returnNumEdgesIter		(explicitly counts each edge)
*/
int returnNumEdgesIter() {
    int numEdges = 0;
    // Iterate over the vertexToEdgeMap incrementing # of edges in each map
    for (auto it = vertexToEdgeMap.begin(); it != vertexToEdgeMap.end(); ++it) {
        numEdges += (it->second).size();
    }

    return numEdges;
}

/*
*	returnNumEdges  		(based only on the counter)
*/
int returnNumEdges() {
    return edgeCounter;
}





















